File: leetcode/0029-divide-two-integers/solution.py
class Solution:
    def divide(self, dividend: int, divisor: int) -> int:
        INT_MIN, INT_MAX = -2**31, 2**31 - 1

        if dividend == INT_MIN and divisor == -1:
            return INT_MAX

        negative = (dividend < 0) != (divisor < 0)

        remaining = abs(dividend)
        div = abs(divisor)

        quotient = 0
        while remaining >= div:
            chunk = div
            multiple = 1
            while remaining >= (chunk << 1):
                chunk <<= 1
                multiple <<= 1
            remaining -= chunk
            quotient += multiple

        if negative:
            quotient = -quotient

        return max(INT_MIN, min(INT_MAX, quotient))

File: leetcode/0029-divide-two-integers/solution.cpp
#include <climits>
#include <cstdlib>
using namespace std;

class Solution {
public:
    int divide(int dividend, int divisor) {
        if (dividend == INT_MIN && divisor == -1) {
            return INT_MAX;
        }

        bool negative = (dividend < 0) != (divisor < 0);

        long long remaining = llabs((long long)dividend);
        long long div = llabs((long long)divisor);

        long long quotient = 0;
        while (remaining >= div) {
            long long chunk = div;
            long long multiple = 1;
            while (remaining >= (chunk << 1)) {
                chunk <<= 1;
                multiple <<= 1;
            }
            remaining -= chunk;
            quotient += multiple;
        }

        if (negative) {
            quotient = -quotient;
        }

        if (quotient > INT_MAX) return INT_MAX;
        if (quotient < INT_MIN) return INT_MIN;
        return (int)quotient;
    }
};
